workflow/scripts/setup_prediction.py

```diff
@@ -0,0 +1,152 @@
+"""
+Set up the VCF prediction pipeline 
+"""
+
+import zarr
+import tskit
+import os
+import yaml
+import numpy as np
+from pysam import FastaFile
+from numcodecs import Blosc
+from math import ceil
+import ts_simulators
+
+
+vcz = zarr.open(snakemake.input.vcz)
+n_variants = vcz.variant_position.size
+n_samples = vcz.sample_id.size
+n_chunks = len(snakemake.output.yaml)
+
+prediction_config = snakemake.params.prediction_config.copy()
+ancestral_fasta_path = prediction_config.get("ancestral_fasta")
+population_map_path = prediction_config.get("population_map")
+#window_type = prediction_config["window_type"]
+#window_size = prediction_config["window_size"]
+
+
+# check the sample configuration in the data against that from the simulator
+simulator_config = snakemake.params.simulator_config.copy()
+simulator = getattr(ts_simulators, simulator_config["class_name"])(simulator_config)
+ts_sim, theta_sim = simulator(seed=1024)
+
+population_id = [p.metadata["name"] for p in ts_sim.populations()]
+population_coding = {p: i for i, p in enumerate(population_id)}
+population_map = yaml.safe_load(open(population_map_path))
+sample_map = {sid: i for i, sid in enumerate(vcz.sample_id)}
+sample_population = np.full(n_samples, -1)
+for ind, pop in population_map.items():
+    assert pop in population_coding, f"Population {pop} not found in simulated instance"
+    assert ind in sample_map, f"Sample {ind} not found in VCF"
+    sample_population[sample_map[ind]] = population_coding[pop]
+    # if a sample from the VCF is not found in the population map, it will be
+    # coded as -1 and will subsequently be omitted when building trees
+
+sample_counts_sim = np.bincount(
+    ts_sim.individuals_population, 
+    minlength=len(population_coding),
+)
+sample_counts = np.bincount(
+    sample_population[sample_population > -1], 
+    minlength=len(population_coding),
+)
+assert np.allclose(sample_counts, sample_counts_sim), \
+    f"Sample counts mismatch between simulation ({sample_counts_sim}) and data ({sample_counts})"
+
+schema = tskit.MetadataSchema.permissive_json()
+populations_metadata = [schema.validate_and_encode_row({"name": p}) for p in population_id]
+zarr.save(f"{snakemake.input.vcz}/individuals_population", sample_population)
+zarr.save(f"{snakemake.input.vcz}/populations_metadata", populations_metadata)
+
+
+# add ancestral state information to vcz
+ancestral_fasta = FastaFile(ancestral_fasta_path)
+ancestral_state = np.full(vcz.variant_position.size, "N", dtype="U1")
+for i, contig in enumerate(vcz.contig_id):
+    fasta = np.fromiter(ancestral_fasta.fetch(reference=contig), dtype="U1")
+    contig_mask = vcz.variant_contig[:] == i
+    position = vcz.variant_position.get_mask_selection(contig_mask)
+    ancestral_state[contig_mask] = np.char.upper(fasta[position - 1])
```

Comment:
i'm worried this could be off by 1 if the vcf came from msprime. we'd want to shift those positions I think.

Reply:
not sure I understand -- the simulated data i'm using here is from msprime, i'm just being careful to generate the right ancestral fasta in the simulations (see the resources/util/simulate-vcf.py script). In general, where data won't come from msprime, I think this is doing the right thing (as VCF indexing is 1-based)